File: vol2/ch19/LeetCode332.py
from collections import defaultdict
class Solution:
    def findItinerary(self,tickets):
        self.hmap=defaultdict(list)				#定义一个字典
        for item in tickets:
            self.hmap[item[0]].append(item[1])
        self.ans=['JFK']
        n=len(tickets)
        self.dfs(n,0)
        return self.ans
    def dfs(self,n,i):
        if i==n:                    #找到一个解
            return True
        else:
            f=self.ans[len(self.ans)-1]     #找到当前出发机场f
            self.hmap[f].sort()             #目的城市递增排序
            for _ in self.hmap[f]:
                t=self.hmap[f].pop(0)  #删除第一个目的城市t
                self.ans.append(t)      #选择第一个目的城市t
                if self.dfs(n,i+1): return True
                self.ans.pop()          #回溯
                self.hmap[f].append(t)
        return False

File: vol2/ch19/LeetCode332.cpp
class Solution {
// unordered_map<出发城市,map<到达城市,航班次数>> hmap
    unordered_map<string,map<string,int>> hmap;
    vector<string> ans;
public:
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        for (auto &ft:tickets)      //构造hmap
            hmap[ft[0]][ft[1]]++;
        ans.push_back("JFK");
        int n=tickets.size();       //n个航线
        dfs(n,0);
        return ans;
    }
    bool dfs(int n,int i) {
        if (i==n)
            return true;
        else {
            string from=ans[ans.size()-1];
            for (auto &to:hmap[from]) {
                if (to.second>0 ) {         //存在<from,to>航线
                    ans.push_back(to.first);
                    to.second--;
                    if (dfs(n,i+1)) return true;
                    ans.pop_back();         //回溯
                    to.second++;
                }
            }
        }
        return false;
    }
};

